### ISSUE TITLE
Can you please check whether the SemiComplexPolicy is correctly implemented?

### ISSUE THREAD

**Issue by m33x** [cupslab/neural_network_cracking#11]
I fear **SemiComplexPolicy** and **SemiComplexPolicyLowercase** need to be patched.
Depending on how one has measured and implemented it, the results for 3class12 in the paper might be unreliable. Please check, whether or not I'm wrong here.

I think you will need a `not` in front of `if self.all_from_group(pwd, self.non_symbols):`

Please have a look here
https://github.com/cupslab/neural_network_cracking/blob/master/pwd_guess.py#L2069
and here
https://github.com/cupslab/neural_network_cracking/blob/master/pwd_guess.py#L2085

---

A better fix would be to stop using **"if not non_symbols"** but **"if symbols"**.

You can generate **symbols** like this
```Python
non_symbols = set(string.digits + string.ascii_uppercase + string.ascii_lowercase)
# ASCII-95 == 0x20 to 0x7E (32 to 126 decimal)
# Note: using string.punctuation + string.whitespace or string.printable is not the same.
# Instead we use the characters as defined here https://en.wikibooks.org/wiki/C%2B%2B_Programming/ASCII
all_95 = set([chr(i) for i in range(32,127)])
symbols = filter(lambda x:x not in non_symbols, all_95)
```
this way you simply stop using the function `all_from_group`and replace it with a standard conform `if has_group(pwd, symbols):`. This would increase readability a lot.

**Comment by wrmelicher** [comment]
I took a quick look at it, and I'm not sure that this is an issue yet. I need to think about it and investigate more. As for the results for 3class12 in the paper, the test set was definitely 3class12 since it was collected from a separate data set which didn't touch this portion of the code. It might be possible that this method of selecting training data isn't ideal though.
